ai_feedback/helpers/template_utils.py

```diff
@@ -64,69 +64,120 @@ def render_prompt_template(
-        filename = os.path.basename(file_path)
+    if submission:
+        file_contents += _format_file_with_xml_tag(submission, "submission")
 
-        try:
-            # Handle PDF files separately
-            if filename.lower().endswith('.pdf'):
-                text_content = extract_pdf_text(file_path)
-                lines = text_content.split('\n')
-            else:
-                # Handle regular text files
-                with open(file_path, "r", encoding="utf-8") as file:
-                    lines = file.readlines()
-
-            # Common processing for both file types
-            file_contents += f"=== {filename} ===\n"
-            for i, line in enumerate(lines, start=1):
-                stripped_line = line.rstrip('\n').rstrip()
-                if stripped_line.strip():
-                    file_contents += f"(Line {i}) {stripped_line}\n"
-                else:
-                    file_contents += f"(Line {i}) \n"
-            file_contents += "\n"
-
-        except Exception as e:
-            print(f"Error reading file {filename}: {e}")
-            continue
+    if solution:
+        file_contents += _format_file_with_xml_tag(solution, "solution")
+
+    if test_output:
+        file_contents += _format_file_with_xml_tag(test_output, "test_output")
 
     return file_contents
 
 
+def _format_file_with_xml_tag(file_path: Path, tag_name: str) -> str:
+    """Format a single file with XML tags and line numbers.
+
+    Args:
+        file_path (Path): Path to the file to format
+        tag_name (str): The XML tag name (submission, solution, test_output)
+
+    Returns:
+        str: Formatted file content with XML tags
+    """
+    if not file_path:
+        return ""
+
+    filename = os.path.basename(file_path)
+
+    try:
+        # Handle PDF files separately
+        if filename.lower().endswith('.pdf'):
+            text_content = extract_pdf_text(file_path)
+            lines = text_content.split('\n')
+            return _wrap_lines_with_xml(lines, tag_name, filename, is_pdf=True)
+        else:
+            # Handle regular text files
+            with open(file_path, "r", encoding="utf-8") as file:
+                lines = file.readlines()
+            return _wrap_lines_with_xml(lines, tag_name, filename, is_pdf=False)
+
+    except Exception as e:
+        print(f"Error reading file {filename}: {e}")
+        return ""
+
+
+def _wrap_lines_with_xml(lines: List[str], tag_name: str, filename: str, is_pdf: bool = False) -> str:
+    """Wrap lines with XML tags and add line numbers.
+
+    Args:
+        lines (List[str]): List of lines to format
+        tag_name (str): The XML tag name (submission, solution, test_output)
+        filename (str): The filename to include in the XML tag
+        is_pdf (bool): Whether this is PDF content (affects empty line handling)
+
+    Returns:
+        str: Formatted content with XML tags and line numbers
+    """
+    content = f"<{tag_name} file=\"{filename}\">\n"
```

Comment:
Use the attribute name `filename` instead of `file`, here and throughout all XML tags that have to do with files